# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/fs/parent/datamodel/FileSystemDataManager.h
# Commit: 0256861a7ac5
# Full Hash: 0256861a7ac5da2076df7735d1522fca9725b346
# Author: Jan Varga <jvarga@mozilla.com>
# Date: 2022-08-31 21:54:47
# Regressor Bug: 1786465
# File Overlap Count: 3
# Description:
#   Bug 1786465 - Implement registration counting of FileSystemDataManager; r=dom-storage-reviewers,jesup
#   
#   FileSystemDataManager can be now closed when it's still alive (not in the
#   destructor).
#   
# ==============================================================================

diff -r 04e60585a3f0 -r 0256861a7ac5 dom/fs/parent/datamodel/FileSystemDataManager.h
--- a/dom/fs/parent/datamodel/FileSystemDataManager.h	Wed Aug 31 09:34:57 2022 +0000
+++ b/dom/fs/parent/datamodel/FileSystemDataManager.h	Wed Aug 31 09:34:57 2022 +0000
@@ -10,25 +10,33 @@
 #include "mozilla/NotNull.h"
 #include "mozilla/TaskQueue.h"
 #include "mozilla/dom/FileSystemTypes.h"
+#include "mozilla/dom/FileSystemHelpers.h"
 #include "mozilla/dom/quota/CheckedUnsafePtr.h"
 #include "nsCOMPtr.h"
 #include "nsISupportsUtils.h"
 #include "nsString.h"
+#include "nsTHashSet.h"
 
 namespace mozilla {
 
 template <typename V, typename E>
 class Result;
 
-namespace dom::fs::data {
+namespace dom {
+
+class FileSystemManagerParent;
+
+namespace fs::data {
 
 class FileSystemDataManager
     : public SupportsCheckedUnsafePtr<CheckIf<DiagnosticAssertEnabled>> {
  public:
+  enum struct State : uint8_t { Initial = 0, Closed };
+
   FileSystemDataManager(const Origin& aOrigin,
                         MovingNotNull<RefPtr<TaskQueue>> aIOTaskQueue);
 
-  using result_t = Result<RefPtr<FileSystemDataManager>, nsresult>;
+  using result_t = Result<Registered<FileSystemDataManager>, nsresult>;
   static FileSystemDataManager::result_t GetOrCreateFileSystemDataManager(
       const Origin& aOrigin);
 
@@ -42,15 +50,31 @@
     return mIOTaskQueue.get();
   }
 
+  void Register();
+
+  void Unregister();
+
+  void RegisterActor(NotNull<FileSystemManagerParent*> aActor);
+
+  void UnregisterActor(NotNull<FileSystemManagerParent*> aActor);
+
  protected:
   ~FileSystemDataManager();
 
+  bool IsInactive() const;
+
+  void Close();
+
+  nsTHashSet<FileSystemManagerParent*> mActors;
   const Origin mOrigin;
   const NotNull<nsCOMPtr<nsISerialEventTarget>> mBackgroundTarget;
   const NotNull<RefPtr<TaskQueue>> mIOTaskQueue;
+  uint32_t mRegCount;
+  State mState;
 };
 
-}  // namespace dom::fs::data
+}  // namespace fs::data
+}  // namespace dom
 }  // namespace mozilla
 
 #endif  // DOM_FS_PARENT_DATAMODEL_FILESYSTEMDATAMANAGER_H_
